test/Marketplace.test.js

```diff
@@ -652,14 +697,41 @@ describe("Marketplace", function () {
         expectedPartialHostPayout
       )
     })
+
+    it("when slot is freed and not filled, client will get refunded the freed slot's funds", async function () {
+      await waitUntilStarted(marketplace, request, proof, token, [slot.index])
+
+      await advanceTimeForNextBlock(request.expiry / 2)
+      await marketplace.fillSlot(requestId(request), slot.index, proof)
+      const id = slotId(slot)
+      const filledAt = await currentTime()
+
+      await expect(marketplace.freeSlot(id)).to.emit(marketplace, "SlotFreed")
+      await waitUntilFinished(marketplace, requestId(request))
+
+      const expectedHostPayout =
+        ((await marketplace.requestEnd(requestId(request))).toNumber() -
+          filledAt) *
+        request.ask.reward
+
+      switchAccount(client)
+      await marketplace.withdrawFunds(slot.request)
+      const endBalance = await token.balanceOf(client.address)
+      expect(
+        maxPrice(request) -
+          (ACCOUNT_STARTING_BALANCE - endBalance) -
+          (request.expiry / 2) * request.ask.reward - // we have moved one slot to mid-expiry window
+          17 * request.ask.reward // There was 17 transaction for filling/advancing etc
```

Comment:
If you took a `startBalance` sample before calling `withdrawFunds`, shouldn't `expectedHostPayout` be equal to `(endBalance - startBalance)`? The idea being that if all four slots were filled for the entire duration, the client should have no refunds, and `(endBalance - startBalance)` would be 0. However, since one slot is freed after starting, and we expect a refund, the difference should be equal to the refund.

Reply:
Hmm actually I think it would not be 0 if there were no refunds as it depends when the slots were filled, and each transaction produces a block that is advanced by 1 second. So maybe instead of having a function `waitUntilStarted` with a parameter to not start certain slots, maybe there could be a function that fills certain slots and returns an array of start times? Those start times could be used to calculate how much each slot should be paid.

Reply:
Yeah, this is something I more or less did. Please check the reworked test.